File: packages/modules/chess/ts/src/perft.ts
import type { GameState } from './types';
import { cloneBoard } from './board';
import { legalMoves, applyMove } from './moves';
import { moveToUCI } from './notation';
import { rankOf, fileOf, square } from './utils';

export const perft = (state: GameState, depth: number): number => {
  if (depth === 0) return 1;

  const moves = legalMoves(
    state.board,
    state.turn,
    state.castlingRights,
    state.enPassant
  );

  if (depth === 1) return moves.length;

  let count = 0;

  for (const move of moves) {
    const newBoard = cloneBoard(state.board);
    applyMove(newBoard, move);

    const newTurn = state.turn === 'w' ? 'b' : 'w';

    const nextRights = { ...state.castlingRights };
    if (move.from === 4 || move.to === 4) nextRights.K = nextRights.Q = false;
    if (move.from === 60 || move.to === 60) nextRights.K = nextRights.Q = false;
    if (move.from === 7 || move.to === 7) nextRights.K = false;
    if (move.from === 0 || move.to === 0) nextRights.Q = false;
    if (move.from === 63 || move.to === 63) nextRights.k = false;
    if (move.from === 56 || move.to === 56) nextRights.q = false;

    const piece = state.board[move.from]!;
    if (piece.type === 'k') {
      if (piece.color === 'w') nextRights.K = nextRights.Q = false;
      else nextRights.k = nextRights.q = false;
    }
    if (piece.type === 'r') {
      if (move.from === 7) nextRights.K = false;
      if (move.from === 0) nextRights.Q = false;
      if (move.from === 63) nextRights.k = false;
      if (move.from === 56) nextRights.q = false;
    }

    const fromRank = rankOf(move.from);
    const toRank = rankOf(move.to);
    const toFile = fileOf(move.to);

    const nextEnPassant =
      piece.type === 'p' && Math.abs(toRank - fromRank) === 2
        ? square((fromRank + toRank) / 2, toFile)
        : null;

    const nextState: GameState = {
      board: newBoard,
      turn: newTurn,
      castlingRights: nextRights,
      enPassant: nextEnPassant,
      halfMoveClock: 0,
      fullMoveNumber: 1,
      history: [],
      status: 'playing',
      result: '*',
      inCheck: false,
    };

    count += perft(nextState, depth - 1);
  }

  return count;
};

export const divide = (
  state: GameState,
  depth: number
): Record<string, number> => {
  const result: Record<string, number> = {};
  const moves = legalMoves(
    state.board,
    state.turn,
    state.castlingRights,
    state.enPassant
  );

  for (const move of moves) {
    const uci = moveToUCI(move);

    const newBoard = cloneBoard(state.board);
    applyMove(newBoard, move);

    const newTurn = state.turn === 'w' ? 'b' : 'w';

    const nextRights = { ...state.castlingRights };
    if (move.from === 4 || move.to === 4) nextRights.K = nextRights.Q = false;
    if (move.from === 60 || move.to === 60) nextRights.K = nextRights.Q = false;
    if (move.from === 7 || move.to === 7) nextRights.K = false;
    if (move.from === 0 || move.to === 0) nextRights.Q = false;
    if (move.from === 63 || move.to === 63) nextRights.k = false;
    if (move.from === 56 || move.to === 56) nextRights.q = false;

    const piece = state.board[move.from]!;
    if (piece.type === 'k') {
      if (piece.color === 'w') nextRights.K = nextRights.Q = false;
      else nextRights.k = nextRights.q = false;
    }
    if (piece.type === 'r') {
      if (move.from === 7) nextRights.K = false;
      if (move.from === 0) nextRights.Q = false;
      if (move.from === 63) nextRights.k = false;
      if (move.from === 56) nextRights.q = false;
    }

    const fromRank = rankOf(move.from);
    const toRank = rankOf(move.to);
    const toFile = fileOf(move.to);

    const nextEnPassant =
      piece.type === 'p' && Math.abs(toRank - fromRank) === 2
        ? square((fromRank + toRank) / 2, toFile)
        : null;

    const nextState: GameState = {
      board: newBoard,
      turn: newTurn,
      castlingRights: nextRights,
      enPassant: nextEnPassant,
      halfMoveClock: 0,
      fullMoveNumber: 1,
      history: [],
      status: 'playing',
      result: '*',
      inCheck: false,
    };

    result[uci] = depth > 1 ? perft(nextState, depth - 1) : 1;
  }

  return result;
};

File: packages/modules/chess/go/core/perft/perft.go
package perft

import (
	"github.com/hieudoanm/chess/core/board"
	"github.com/hieudoanm/chess/core/moves"
	"github.com/hieudoanm/chess/core/types"
	"github.com/hieudoanm/chess/core/utils"
)

func Perft(state types.GameState, depth int) int {
	if depth == 0 {
		return 1
	}

	movesList := moves.LegalMoves(state.Board, state.Turn, state.CastlingRights, state.EnPassant)

	if depth == 1 {
		return len(movesList)
	}

	count := 0

	for _, move := range movesList {
		newBoard := board.CloneBoard(state.Board)
		moves.ApplyMove(&newBoard, move)

		newTurn := state.Turn.Opposite()

		nextRights := state.CastlingRights
		if move.From == 4 || move.To == 4 {
			nextRights.WK = false
			nextRights.WQ = false
		}
		if move.From == 60 || move.To == 60 {
			nextRights.WK = false
			nextRights.WQ = false
		}
		if move.From == 7 || move.To == 7 {
			nextRights.WK = false
		}
		if move.From == 0 || move.To == 0 {
			nextRights.WQ = false
		}
		if move.From == 63 || move.To == 63 {
			nextRights.BK = false
		}
		if move.From == 56 || move.To == 56 {
			nextRights.BQ = false
		}

		piece := state.Board[move.From]
		if piece != nil {
			if piece.Type == types.King {
				if piece.Color == types.White {
					nextRights.WK = false
					nextRights.WQ = false
				} else {
					nextRights.BK = false
					nextRights.BQ = false
				}
			}
			if piece.Type == types.Rook {
				if move.From == 7 {
					nextRights.WK = false
				}
				if move.From == 0 {
					nextRights.WQ = false
				}
				if move.From == 63 {
					nextRights.BK = false
				}
				if move.From == 56 {
					nextRights.BQ = false
				}
			}
		}

		fromRank := utils.RankOf(move.From)
		toRank := utils.RankOf(move.To)
		toFile := utils.FileOf(move.To)

		nextEnPassant := types.NoSquare
		if piece != nil && piece.Type == types.Pawn && abs(toRank-fromRank) == 2 {
			nextEnPassant = utils.Square((fromRank+toRank)/2, toFile)
		}

		nextState := types.GameState{
			Board:          newBoard,
			Turn:           newTurn,
			CastlingRights: nextRights,
			EnPassant:      nextEnPassant,
			HalfMoveClock:  0,
			FullMoveNumber: 1,
			History:        nil,
			Status:         types.StatusPlaying,
			Result:         types.ResultOngoing,
			InCheck:        false,
		}

		count += Perft(nextState, depth-1)
	}

	return count
}

func Divide(state types.GameState, depth int) map[string]int {
	result := make(map[string]int)
	movesList := moves.LegalMoves(state.Board, state.Turn, state.CastlingRights, state.EnPassant)

	for _, move := range movesList {
		uci := notationMoveToUCI(move)

		newBoard := board.CloneBoard(state.Board)
		moves.ApplyMove(&newBoard, move)

		newTurn := state.Turn.Opposite()

		nextRights := state.CastlingRights
		if move.From == 4 || move.To == 4 {
			nextRights.WK = false
			nextRights.WQ = false
		}
		if move.From == 60 || move.To == 60 {
			nextRights.WK = false
			nextRights.WQ = false
		}
		if move.From == 7 || move.To == 7 {
			nextRights.WK = false
		}
		if move.From == 0 || move.To == 0 {
			nextRights.WQ = false
		}
		if move.From == 63 || move.To == 63 {
			nextRights.BK = false
		}
		if move.From == 56 || move.To == 56 {
			nextRights.BQ = false
		}

		piece := state.Board[move.From]
		if piece != nil {
			if piece.Type == types.King {
				if piece.Color == types.White {
					nextRights.WK = false
					nextRights.WQ = false
				} else {
					nextRights.BK = false
					nextRights.BQ = false
				}
			}
			if piece.Type == types.Rook {
				if move.From == 7 {
					nextRights.WK = false
				}
				if move.From == 0 {
					nextRights.WQ = false
				}
				if move.From == 63 {
					nextRights.BK = false
				}
				if move.From == 56 {
					nextRights.BQ = false
				}
			}
		}

		fromRank := utils.RankOf(move.From)
		toRank := utils.RankOf(move.To)
		toFile := utils.FileOf(move.To)

		nextEnPassant := types.NoSquare
		if piece != nil && piece.Type == types.Pawn && abs(toRank-fromRank) == 2 {
			nextEnPassant = utils.Square((fromRank+toRank)/2, toFile)
		}

		nextState := types.GameState{
			Board:          newBoard,
			Turn:           newTurn,
			CastlingRights: nextRights,
			EnPassant:      nextEnPassant,
			HalfMoveClock:  0,
			FullMoveNumber: 1,
			History:        nil,
			Status:         types.StatusPlaying,
			Result:         types.ResultOngoing,
			InCheck:        false,
		}

		if depth > 1 {
			result[uci] = Perft(nextState, depth-1)
		} else {
			result[uci] = 1
		}
	}

	return result
}

func notationMoveToUCI(move types.Move) string {
	return utils.SquareName(move.From) + utils.SquareName(move.To) + string(move.Promotion)
}

func abs(n int) int {
	if n < 0 {
		return -n
	}
	return n
}
